profiling/src/collections/string_storage.rs

```diff
@@ -23,6 +22,40 @@ pub struct ManagedStringStorage {
     id_to_data: HashMap<u32, ManagedStringData, BuildHasherDefault<rustc_hash::FxHasher>>,
     str_to_id: HashMap<Rc<str>, u32, BuildHasherDefault<rustc_hash::FxHasher>>,
     current_gen: u32,
+    next_cached_profile_id: InternalCachedProfileId,
+}
+
+#[derive(PartialEq, Debug)]
+// The `ManagedStringStorage::get_seq_num` operation is used to map a `ManagedStorageId` into a
+// `StringId` for a given `StringTable`. As an optimization, we store a one-element `cached_seq_num`
+//  inline cache with each `ManagedStringData` entry, so that repeatedly calling
+// `get_seq_num` with the same id provides faster lookups.
+//
+// Because:
+// 1. Multiple profiles can be using the same `ManagedStringTable`
+// 2. The same profile resets its `StringTable` on serialization and starts anew
+// ...we need a way to identify when the cache can and cannot be reused.
+//
+// This is where the `CachedProfileId` comes in. A given `CachedProfileId` should be considered
+// as representing a unique `StringTable`. Different `StringTable`s should have different
+// `CachedProfileId`s, and when a `StringTable` gets flushed and starts anew, it should also have a
+// different `CachedProfileId`.
+//
+// **This struct is on purpose not Copy and not Clone to try to make it really hard to accidentally
+// reuse** when a profile gets reset.
+pub struct CachedProfileId {
+    id: u32,
+}
+
+#[derive(PartialEq, Debug, Copy, Clone)]
+struct InternalCachedProfileId {
+    id: u32,
```

Comment:
Do we want to use an incrementing number? If there are two profiles we switch between, would we get conflicts on this id?